### ISSUE TITLE
USB support

### ISSUE THREAD

**Issue by josephernest** [yaq-project/yaqd-horiba#33]
Hello, I read this in your readme:

> This program does not use the manufacturer provided drivers for the device. Instead it uses libusb. I use https://zadig.akeo.ie/ to override the driver to WinUSB. (This has the advantage of being able to be rolled back to the manufacturer drivers if needed)

Is it possible to talk to the iHR320 *without* using a custom driver (Zadig, etc.)?

For example can I just move the wavelength with `pyserial` (19200 bauds, etc.), or simply with `pyusb` which usually works well? (pip install pyusb). 

Can we avoid to install Zadig etc.?

**Comment by untzag** [comment]
Hi @josephernest, great questions. This project mostly belongs to @ksunden, but I'll give my thoughts here.

The horiba interface is not serial or anything resembling UART. I don't see any way for us to use `pyserial`. Horiba's own software sends USB packets, so we have to do the same.

We do use `pyusb`, I'm glad you're a fan! Unfortunately, `pyusb` cannot interface via Horiba's own drivers---Horiba just never intended for other applications to take that route. This is why we've ended up swapping to generic open source drivers.

A long time ago we used to make DLL calls into Horiba's libraries, but those DLLs were not very stable and we really wanted cross platform performance.

That's the story as I know it! If you have other ideas we would love to hear them---always looking to improve. Here's the file where we actually send USB packets if you're curious.

https://github.com/yaq-project/yaqd-horiba/blob/main/yaqd_horiba/_horiba_mono.py

**Comment by josephernest** [comment]
Thanks @untzag and @ksunden for your answers!

Do you think there is a way, on Windows, to keep the original manufacturer driver? My colleagues use the official Horiba software, so I don't want to swap between original and Zadig drivers all the time :)

It nearly works with the original official driver:

```
import usb.core, usb.backend.libusb1
backend = usb.backend.libusb1.get_backend(find_library=lambda x: r"C:\Python38\Lib\site-packages\libusb\_platform\_windows\x64\libusb-1.0.dll")
print(backend)
JOBIN_YVON_ID_VENDOR = 0xC9B
IHR320_ID_PRODUCT = 0x101
dev = usb.core.find(idVendor=JOBIN_YVON_ID_VENDOR, idProduct=IHR320_ID_PRODUCT)
print(dev)
print(dev.serial_number)
```

```
<usb.backend.libusb1._LibUSB object at 0x0000000002B238E0>
DEVICE ID ----:---- on Bus 002 Address 003 =================
 bLength                :   0x12 (18 bytes)
 bDescriptorType        :    0x1 Device
 bcdUSB                 :  0x200 USB 2.0
 bDeviceClass           :    0x0 Specified at interface
 bDeviceSubClass        :    0x0
 bDeviceProtocol        :    0x0
 bMaxPacketSize0        :   0x40 (64 bytes)
 idVendor               : 0x0c9b
 idProduct              : 0x0101
 bcdDevice              :  0x100 Device 1.0
 iManufacturer          :    0x1 Error Accessing String
 iProduct               :    0x2 Error Accessing String
 iSerialNumber          :    0x3 Error Accessing String
 bNumConfigurations     :    0x1
  CONFIGURATION 1: 100 mA ==================================
   bLength              :    0x9 (9 bytes)
   bDescriptorType      :    0x2 Configuration
   wTotalLength         :   0x20 (32 bytes)
   bNumInterfaces       :    0x1
```

except the last line

    print(dev.serial_number)

which gives

    NotImplementedError: Operation not supported or unimplemented on this platform

Do you see a way with `libusb` + keeping the official original driver?

**Comment by ksunden** [comment]
If you got that far without swapping out the driver, it may actually be possible... 

We had to do this workaround even for the libusb drivers:

https://github.com/yaq-project/yaqd-horiba/blob/8a0d400179079f73c91066bfd7da2c5eb5c2df97/yaqd_horiba/_horiba_mono.py#L54-L56

Which I think was the addressing same error message.

**Comment by josephernest** [comment]
Oh this would be cool @ksunden if we could keep the original driver.
Unfortunately, setting `_langids` doesn't help:

```
import usb.core, usb.backend.libusb1
backend = usb.backend.libusb1.get_backend(find_library=lambda x: r"C:\Python38\Lib\site-packages\libusb\_platform\_windows\x64\libusb-1.0.dll")
print(backend)
JOBIN_YVON_ID_VENDOR = 0xC9B
IHR320_ID_PRODUCT = 0x101
LANG_ID_US_ENGLISH = 0x409
dev = usb.core.find(idVendor=JOBIN_YVON_ID_VENDOR, idProduct=IHR320_ID_PRODUCT)
dev._langids = (LANG_ID_US_ENGLISH,)
print(dev)
print(dev.serial_number)  
# NotImplementedError: Operation not supported or unimplemented on this platform
dev.ctrl_transfer(0x40, 0xB3, wValue=0, wIndex=0)
# NotImplementedError: Operation not supported or unimplemented on this platform
```

Maybe with this driver, we have to use another way to set the lang attribute than

    dev._langids = (LANG_ID_US_ENGLISH,)

?

**Comment by ksunden** [comment]
Possibly try different libusb dlls? `libusb-0.1`, `libusbK`, or `WinUSB` come to mind, if you can locate those `dll`s

It's unfortunately a bit of a mess and hard to tell what will work without just trying it.

The [libusb gh wiki](https://github.com/libusb/libusb/wiki/FAQ#libusb-win32-libusbk-and-libusb-project-what-about-openusb) can provide a bit more context than I can.


I will note that we explicitly prefer `libusb0` in the code, so that may be more likely to work:

https://github.com/yaq-project/yaqd-horiba/blob/8a0d400179079f73c91066bfd7da2c5eb5c2df97/yaqd_horiba/_horiba_mono.py#L41-L50

**Comment by ksunden** [comment]
Should be something like `libusb-0.1.dll`, possibly `libusb-compat-0.1.dll`, which is the compatibility layer that actually uses libusb 1.0 under the hood, but is mostly there...

Unfortunately, I don't have a windows box to test with (nor immediate access to the particular hardware). so to a degree I'm guessing here.

That said, it is looking more like that error is what libusb shows when the drivers are not what it expects, while it can get some info (which is more than I remember getting when I tried...) it may not be able to proceed...:

https://stackoverflow.com/questions/17350177/libusb-open-returns-libusb-error-not-supported-on-windows-7

(See here, where the accepted answer states as much and the top voted answer is to use Zadig (from the author of Zadig)...)

So, looking a little less likely than I was hopeful for when I saw that you could at least _see_ the device.

**Comment by josephernest** [comment]
@ksunden @untzag I'm about to rework on the IHR320 + Python. 

If I use `libusb` + https://zadig.akeo.ie/ to override the driver to WinUSB, how can I still use the official software for IHR320 (shipped by Horiba) which won't work with this alternate driver?

Each time I want to go from one software (official Horiba) to another (Python code), do I need to uninstall the previous driver and reinstall the other driver (official vs. libusb/zadig)?

**Comment by ksunden** [comment]
Yes, unfortunately the drivers are not compatible with each other.

It is relatively easy to revert to the manufacturer drivers (I don't remember the exact button sequence, but essentially there is an "undo" button in the device settings for the driver. (And then using zadig you can "redo")

If this is not a _common_ part of your workflow this is workable, though not ideal... It really would be annoying to do regularly though, I acknowledge.

If there are features that we could add to make it so you don't have to switch back and forth, I'm open to considering them (though admittedly my own access to the hardware is limited these days)

**Comment by untzag** [comment]
@ksunden do you know if there is a programmatic way to switch drivers? Does Zadig have an API?

I could imagine something like a context manager for driver loading.

**Comment by ksunden** [comment]
I suspect it is possible, if we have the `.inf` file for each case to use `pnputil` (or `devcon`, but that seems to be deprecated):

https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/devcon-migration

A few hurdles would have to be surpassed:

- Identifying the device
- Knowing where the drivers are (possibly shipping libusb ones? Not enthused about shipping binaries ourselves, but may be better than trying to either locate it on arbitrary machines or downloading and installing otherwise)
- Issuing the commands
- Cleaning up after ourselves (preferably in a robust way)
- May have hard to identify/troubleshoot problems if devices are in use when you try to do something that rewrites drivers.


Zadig takes longer than I would like/expect, so not sure what it is doing other than essentially calling the correct "install driver" incantation.

All in all, I think it may be worth pursuing, but wish I had somebody more proficient in Windows (at the driver level) to tell us if the idea is flawed.

**Comment by ksunden** [comment]
We may also need to contend with requiring higher permissions to execute, which is not ideal... It's one thing to say "do this setup task once as admin" and another to say "every time, run as admin"
